pages/tutorials/3.0/window/events.md

```diff
@@ -7,259 +7,329 @@
 
 ## Introduction
 
-This tutorial is a detailed list of window events. It describes them, and shows how to (and how not to) use them.
+This tutorial is a detailed list of window events.
+It describes them, and shows how to (and how not to) use them.
 
 ## The sf::Event type
 
-Before dealing with events, it is important to understand what the [`sf::Event`](https://www.sfml-dev.org/documentation/3.0.0/classsf_1_1Event.php "sf::Event documentation") type is, and how to correctly use it. [`sf::Event`](https://www.sfml-dev.org/documentation/3.0.0/classsf_1_1Event.php "sf::Event documentation") is a _union_, which means that only one of its members is valid at a time (remember your C++ lesson: all the members of a union share the same memory space). The valid member is the one that matches the event type, for example `event.key` for a `KeyPressed` event. Trying to read any other member will result in an undefined behavior (most likely: random or invalid values). It is important to never try to use an event member that doesn't match its type.
+SFML uses `std::variant` under the hood to implement a type-safe API for events.
+There are two main ways to use this API.
+Check out the EventHandling example program to see these methods in practice.
 
-[`sf::Event`](https://www.sfml-dev.org/documentation/3.0.0/classsf_1_1Event.php "sf::Event documentation") instances are filled by the `pollEvent` (or `waitEvent`) function of the [`sf::Window`](https://www.sfml-dev.org/documentation/3.0.0/classsf_1_1Window.php "sf::Window documentation") class. Only these two functions can produce valid events, any attempt to use an [`sf::Event`](https://www.sfml-dev.org/documentation/3.0.0/classsf_1_1Event.php "sf::Event documentation") which was not returned by successful call to `pollEvent` (or `waitEvent`) will result in the same undefined behavior that was mentioned above.
+### `sf::Event::getIf<T>`
 
-To be clear, here is what a typical event loop looks like:
+The first option is based around `sf::Event::getIf<T>` and `sf::Event::is<T>`.
+`getIf<T>` works by providing a template parameter which must be an event subtype.
+Event subtypes are types like `sf::Event::Resized` or `sf::Event::MouseMoved`.
+If the template argument matches the active event subtype then a pointer to the data is returned.
+If that template argument is not the active event subtype then `nullptr` is returned.
+`sf::Event::is<T>` more simply returns `true` if `T` matches the active event subtype.
+`is<T>` is often used for subtypes like `sf::Event::Closed` which contain no data.
+Here's what that looks like:
```

Comment:
I find the text rather hard to follow. Even though I know the new API, I had to re-read the fourth line three times before I understood everything.

What do you think of the following approach?

```suggestion
The first option is based around `sf::Event`'s member functions `getIf<T>` and `is<T>`.
The template type `T` is an event subtype, such as `sf::Event::Resized` or `sf::Event::MouseMoved`.

`getIf<T>` returns a pointer to the subtype specific event data, if the provided template type matches the active event subtype, otherwise `nullptr` is returned.

`is<T>` returns `true` if the provided template type matches the active event subtype, `false` otherwise.
This is especially useful for subtypes like `sf::Event::Closed` which contain no data.
Here's what that looks like:
```